### ISSUE TITLE
[BUG] VM with assigned PCI GPU can not start after v1.3.2 to v1.4.0-rc4 upgrade due to no /dev/vfio/xx path 

### ISSUE THREAD

**Issue by albinsun** [harvester/harvester#6892]
**Describe the bug**
VM with assigned PCI GPU can not start after v1.3.2 to v1.4.0-rc4 upgrade due to `Error: failed to generate container "xxx" spec: failed to generate spec: lstat /dev/vfio/37: no such file or directory`

![image](https://github.com/user-attachments/assets/2f648587-06d9-46e8-8344-915e193813b8)

**To Reproduce**
1. Setup single node `harvester-v1.3.2`
2. <details>
   <summary> Create a SLES-15.6 VM with a binding PCI passthrough GPU
   </summary>
   
   ![image](https://github.com/user-attachments/assets/210bfca5-96bd-48d3-98a1-d890e6a707ea)
   </details>
   
3. <details>
   <summary> Verify GPU works
   </summary>
   
   ![image](https://github.com/user-attachments/assets/000f151b-71d5-4802-90fa-99c8889241bc)
   </details>
   
4. <details>
   <summary> Upgrade Harvester to <code>v1.4.0-rc4</code>
   </summary>
   
   ![image](https://github.com/user-attachments/assets/5d902053-e23e-40a0-8c8a-6646231dc492)
   </details>

5. :x: Start VM _pci-gpu_ and verify GPU still works
   * VM
     > Error: failed to generate container "3a684b0dee8db2a70f6619728e21f51a5a8aa67dde74fab2eaf32998e02f5ab0" spec: failed to generate spec: lstat /dev/vfio/37: no such file or directory
     ![image](https://github.com/user-attachments/assets/2f648587-06d9-46e8-8344-915e193813b8)
   * <i>virt-launcher</i>
     ![image](https://github.com/user-attachments/assets/8c3d28f3-334a-4a2d-9c4c-8674a60d1b22)

**Expected behavior**
VM with PCI device should still be assigned and functioning after upgrade

**Support bundle**
* upgrade log: [hvst-upgrade-cck7x-upgradelog-archive-2024-10-27T06-41-17Z.zip](https://github.com/user-attachments/files/17533226/hvst-upgrade-cck7x-upgradelog-archive-2024-10-27T06-41-17Z.zip)

* support bundle
  * [supportbundle_AfterUpgrade1_2024-10-27T07-34-13Z.zip](https://github.com/user-attachments/files/17533225/supportbundle_AfterUpgrade1_2024-10-27T07-34-13Z.zip)
  * [supportbundle_AfterUpgrade2_2024-10-27T08-08-38Z.zip](https://github.com/user-attachments/files/17533227/supportbundle_AfterUpgrade2_2024-10-27T08-08-38Z.zip)

**Environment**
 - Harvester ISO version: `v1.3.2` => `v1.4.0-rc4`
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): Single node Bare-metal HPE ProLiant DL360

**Additional context**
1. Other SLES-15.6 based VMs does not hit this problem
   ![image](https://github.com/user-attachments/assets/2fd6bc28-09b8-48fc-ada8-0e3367b8dd33)

**Comment by WebberHuang1118** [comment]
The PCIDevice manifest is:
  ```
  apiVersion: devices.harvesterhci.io/v1beta1
  kind: PCIDevice
  metadata:
    annotations:
      sim.harvesterhci.io/creationTimestamp: "2024-10-25T07:25:30Z"
    creationTimestamp: "2024-10-25T07:25:30Z"
    generation: 1
    labels:
      nodename: qa-191
    managedFields:
    - apiVersion: devices.harvesterhci.io/v1beta1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            f:harvesterhci.io/pcideviceDriver: {}
      manager: pcidevices
      operation: Update
      time: "2024-10-25T07:25:30Z"
    - apiVersion: devices.harvesterhci.io/v1beta1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:sim.harvesterhci.io/creationTimestamp: {}
          f:labels:
            .: {}
            f:nodename: {}
        f:spec: {}
      manager: support-bundle-kit-amd64
      operation: Update
      time: "2024-10-30T09:11:04Z"
    - apiVersion: devices.harvesterhci.io/v1beta1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          .: {}
          f:address: {}
          f:classId: {}
          f:description: {}
          f:deviceId: {}
          f:iommuGroup: {}
          f:kernelDriverInUse: {}
          f:nodeName: {}
          f:resourceName: {}
          f:vendorId: {}
      manager: support-bundle-kit-amd64
      operation: Update
      subresource: status
      time: "2024-10-30T09:11:05Z"
    name: qa-191-000008000
    resourceVersion: "698"
    uid: 5aae9852-6d19-440a-bcec-b5f216ea359f
  spec: {}
  status:
    address: "0000:08:00.0"
    classId: "0302"
    description: '3D controller: NVIDIA Corporation GA107GL [A2 / A16]'
    deviceId: 25b6
    iommuGroup: "32"
    kernelDriverInUse: vfio-pci
    nodeName: qa-191
    resourceName: nvidia.com/GA107GL_A2___A16
    vendorId: 10de
  ```

The `iommuGroup` is `32`,  [the container should access path `/dev/vfio/32`](https://github.com/harvester/pcidevices/blob/c65a0933fd358d7fb631bc62847a5041418ff8f0/pkg/deviceplugins/common.go#L168-L173), however, the container tried to use `/dev/vfio/37` which is not matched to current status.

Following is my inference:
- After rebooting, the PCIDevice's iommGroup changed
- PCIDevice controller [found the current iommuGroupMap](https://github.com/harvester/pcidevices/blob/c65a0933fd358d7fb631bc62847a5041418ff8f0/pkg/controller/pcidevice/pcidevice_controller.go#L48-L52)
- PCIDevice controller [updates iommuGroup number to PCIDevice](https://github.com/harvester/pcidevices/blob/c65a0933fd358d7fb631bc62847a5041418ff8f0/pkg/controller/pcidevice/pcidevice_controller.go#L98-L103)
- However, before the above operation happens, the device plugin already started with the old iommuGroup number

**Comment by Yu-Jack** [comment]
[Update]

After retrying, nothing happens. I'm not sure why or what conditions it shows these errors.

---

In my case, it shows:

```
│ harvester-pcidevices-controller-7x25t time="2024-10-31T07:55:00Z" level=info msg="Enabling passthrough for PDC: node-000002010"                                                                                                    │
│ harvester-pcidevices-controller-7x25t time="2024-10-31T07:55:00Z" level=info msg="Binding device node-000002010 [8086 7ae0] to vfio-pci"                                                                                           │
│ harvester-pcidevices-controller-7x25t time="2024-10-31T07:55:00Z" level=error msg="error syncing 'node-000002010': handler PCIDeviceClaimReconcile: error opening new_id file: open /sys/bus/pci/drivers/vfio-pci/new_id: no such  │
```

This is my `pcidevice` (there is no `kernelDriverInUse` field) and `pcideviceclaim`:

```yaml
apiVersion: devices.harvesterhci.io/v1beta1
kind: PCIDevice
metadata:
  annotations:
    harvesterhci.io/pcideviceDriver: xhci_hcd
  creationTimestamp: "2024-10-31T06:47:06Z"
  generation: 1
  labels:
    nodename: node
  name: node-000002010
  resourceVersion: "40347"
  uid: fc83fd7d-b1f2-40d4-a9cc-34a230315924
spec: {}
status:
  address: "0000:02:01.0"
  classId: 0C03
  description: 'Class 0C03: Intel Corporation Alder Lake-S PCH USB 3.2 Gen 2x2 XHCI
    Controller'
  deviceId: 7ae0
  iommuGroup: "18"
  nodeName: node
  resourceName: intel.com/ALDER_LAKES_PCH_USB_3_2_GEN_2X2_XHCI_CONTROLLER
  vendorId: "8086"
```

```yaml
apiVersion: devices.harvesterhci.io/v1beta1
kind: PCIDeviceClaim
metadata:
  creationTimestamp: "2024-10-31T06:48:02Z"
  finalizers:
  - wrangler.cattle.io/PCIDeviceClaimOnRemove
  generation: 1
  name: node-000002010
  ownerReferences:
  - apiVersion: devices.harvesterhci.io/v1beta1
    kind: PCIDevice
    name: node-000002010
    uid: fc83fd7d-b1f2-40d4-a9cc-34a230315924
  resourceVersion: "21380"
  uid: 42189b95-f824-4ce5-b9ce-09970b0ca39e
spec:
  address: "0000:02:01.0"
  nodeName: node
  userName: admin
status:
  kernelDriverToUnbind: xhci_hcd
  passthroughEnabled: true
```

But, there is no `vfio-pci` folder:

```
harvester-pcidevices-controller-7x25t:/sys/bus/pci/drivers # ls -al
total 0
drwxr-xr-x 28 root root 0 Oct 31 07:07 .
drwxr-xr-x  5 root root 0 Oct 31 07:07 ..
drwxr-xr-x  2 root root 0 Oct 31 07:07 8250_lpss
drwxr-xr-x  2 root root 0 Oct 31 07:07 8250_mid
drwxr-xr-x  2 root root 0 Oct 31 07:07 agpgart-intel
drwxr-xr-x  2 root root 0 Oct 31 07:07 agpgart-sis
drwxr-xr-x  2 root root 0 Oct 31 07:07 agpgart-via
drwxr-xr-x  2 root root 0 Oct 31 07:07 ahci
drwxr-xr-x  2 root root 0 Oct 31 07:07 be2iscsi
drwxr-xr-x  2 root root 0 Oct 31 07:07 cxgb3
drwxr-xr-x  2 root root 0 Oct 31 07:07 cxgb4
drwxr-xr-x  2 root root 0 Oct 31 07:07 dw_dmac_pci
drwxr-xr-x  2 root root 0 Oct 31 07:07 exar_serial
drwxr-xr-x  2 root root 0 Oct 31 07:07 i801_smbus
drwxr-xr-x  2 root root 0 Oct 31 07:07 intel_scu
drwxr-xr-x  2 root root 0 Oct 31 07:07 iosf_mbi_pci
drwxr-xr-x  2 root root 0 Oct 31 07:07 lpc_ich
drwxr-xr-x  2 root root 0 Oct 31 07:07 pci-pf-stub
drwxr-xr-x  2 root root 0 Oct 31 07:07 pci-stub
drwxr-xr-x  2 root root 0 Oct 31 07:07 pcieport
drwxr-xr-x  2 root root 0 Oct 31 07:07 qla4xxx
drwxr-xr-x  2 root root 0 Oct 31 07:07 serial
drwxr-xr-x  2 root root 0 Oct 31 07:07 shpchp
drwxr-xr-x  2 root root 0 Oct 31 07:07 thunderbolt
drwxr-xr-x  2 root root 0 Oct 31 07:07 virtio-pci
drwxr-xr-x  2 root root 0 Oct 31 07:07 vmw_vmci
drwxr-xr-x  2 root root 0 Oct 31 07:07 xen-platform-pci
drwxr-xr-x  2 root root 0 Oct 31 07:07 xhci_hcd
harvester-pcidevices-controller-7x25t:/sys/bus/pci/drivers #
```

**Comment by WebberHuang1118** [comment]
Hi @albinsun 

Based on my findings,
Can you try the following steps, after the issue occurs?
- Stopping the VM
- Disabling the PCIDeviceClaim
- Waiting until the PCIDeviceClaim is removed
- Enabling the PCIDeviceClaim
- Restarting the VM

Thanks

**Comment by WebberHuang1118** [comment]
> [Update]
> 
> After retrying, nothing happens. I'm not sure why or what conditions it shows these errors.
> 
> In my case, it shows:
> 
> ```
> │ harvester-pcidevices-controller-7x25t time="2024-10-31T07:55:00Z" level=info msg="Enabling passthrough for PDC: node-000002010"                                                                                                    │
> │ harvester-pcidevices-controller-7x25t time="2024-10-31T07:55:00Z" level=info msg="Binding device node-000002010 [8086 7ae0] to vfio-pci"                                                                                           │
> │ harvester-pcidevices-controller-7x25t time="2024-10-31T07:55:00Z" level=error msg="error syncing 'node-000002010': handler PCIDeviceClaimReconcile: error opening new_id file: open /sys/bus/pci/drivers/vfio-pci/new_id: no such  │
> ```
> 
> This is my `pcidevice` (there is no `kernelDriverInUse` field) and `pcideviceclaim`:
> 
> ```yaml
> apiVersion: devices.harvesterhci.io/v1beta1
> kind: PCIDevice
> metadata:
>   annotations:
>     harvesterhci.io/pcideviceDriver: xhci_hcd
>   creationTimestamp: "2024-10-31T06:47:06Z"
>   generation: 1
>   labels:
>     nodename: node
>   name: node-000002010
>   resourceVersion: "40347"
>   uid: fc83fd7d-b1f2-40d4-a9cc-34a230315924
> spec: {}
> status:
>   address: "0000:02:01.0"
>   classId: 0C03
>   description: 'Class 0C03: Intel Corporation Alder Lake-S PCH USB 3.2 Gen 2x2 XHCI
>     Controller'
>   deviceId: 7ae0
>   iommuGroup: "18"
>   nodeName: node
>   resourceName: intel.com/ALDER_LAKES_PCH_USB_3_2_GEN_2X2_XHCI_CONTROLLER
>   vendorId: "8086"
> ```
> 
> ```yaml
> apiVersion: devices.harvesterhci.io/v1beta1
> kind: PCIDeviceClaim
> metadata:
>   creationTimestamp: "2024-10-31T06:48:02Z"
>   finalizers:
>   - wrangler.cattle.io/PCIDeviceClaimOnRemove
>   generation: 1
>   name: node-000002010
>   ownerReferences:
>   - apiVersion: devices.harvesterhci.io/v1beta1
>     kind: PCIDevice
>     name: node-000002010
>     uid: fc83fd7d-b1f2-40d4-a9cc-34a230315924
>   resourceVersion: "21380"
>   uid: 42189b95-f824-4ce5-b9ce-09970b0ca39e
> spec:
>   address: "0000:02:01.0"
>   nodeName: node
>   userName: admin
> status:
>   kernelDriverToUnbind: xhci_hcd
>   passthroughEnabled: true
> ```
> 
> But, there is no `vfio-pci` folder:
> 
> ```
> harvester-pcidevices-controller-7x25t:/sys/bus/pci/drivers # ls -al
> total 0
> drwxr-xr-x 28 root root 0 Oct 31 07:07 .
> drwxr-xr-x  5 root root 0 Oct 31 07:07 ..
> drwxr-xr-x  2 root root 0 Oct 31 07:07 8250_lpss
> drwxr-xr-x  2 root root 0 Oct 31 07:07 8250_mid
> drwxr-xr-x  2 root root 0 Oct 31 07:07 agpgart-intel
> drwxr-xr-x  2 root root 0 Oct 31 07:07 agpgart-sis
> drwxr-xr-x  2 root root 0 Oct 31 07:07 agpgart-via
> drwxr-xr-x  2 root root 0 Oct 31 07:07 ahci
> drwxr-xr-x  2 root root 0 Oct 31 07:07 be2iscsi
> drwxr-xr-x  2 root root 0 Oct 31 07:07 cxgb3
> drwxr-xr-x  2 root root 0 Oct 31 07:07 cxgb4
> drwxr-xr-x  2 root root 0 Oct 31 07:07 dw_dmac_pci
> drwxr-xr-x  2 root root 0 Oct 31 07:07 exar_serial
> drwxr-xr-x  2 root root 0 Oct 31 07:07 i801_smbus
> drwxr-xr-x  2 root root 0 Oct 31 07:07 intel_scu
> drwxr-xr-x  2 root root 0 Oct 31 07:07 iosf_mbi_pci
> drwxr-xr-x  2 root root 0 Oct 31 07:07 lpc_ich
> drwxr-xr-x  2 root root 0 Oct 31 07:07 pci-pf-stub
> drwxr-xr-x  2 root root 0 Oct 31 07:07 pci-stub
> drwxr-xr-x  2 root root 0 Oct 31 07:07 pcieport
> drwxr-xr-x  2 root root 0 Oct 31 07:07 qla4xxx
> drwxr-xr-x  2 root root 0 Oct 31 07:07 serial
> drwxr-xr-x  2 root root 0 Oct 31 07:07 shpchp
> drwxr-xr-x  2 root root 0 Oct 31 07:07 thunderbolt
> drwxr-xr-x  2 root root 0 Oct 31 07:07 virtio-pci
> drwxr-xr-x  2 root root 0 Oct 31 07:07 vmw_vmci
> drwxr-xr-x  2 root root 0 Oct 31 07:07 xen-platform-pci
> drwxr-xr-x  2 root root 0 Oct 31 07:07 xhci_hcd
> harvester-pcidevices-controller-7x25t:/sys/bus/pci/drivers #
> ```

Hi @Yu-Jack ,
I think your case is different,
in the situation found by @albinsun, the vfio driver seems already bound, but the iommuGroup is not matched during create the VM container.

**Comment by WebberHuang1118** [comment]
> Hi @albinsun
> 
> Based on my findings, Can you try the following steps, after the issue occurs?
> 
> * Stopping the VM
> * Disabling the PCIDeviceClaim
> * Waiting until the PCIDeviceClaim is removed
> * Enabling the PCIDeviceClaim
> * Restarting the VM
> 
> Thanks

The VM can restart after applying the above steps,
https://github.com/harvester/harvester/issues/6892#issuecomment-2446509516 looks like the root cause.

**Comment by albinsun** [comment]
Summarize test result:
1. Retest still hit with following steps:
   1. Single node `v1.3.2` with a running VM which has a GPU PCI-Passthrough
   2. Upgrade to `v1.4.0-rc4`
   3. After upgrade, the VM becomes `Off`
   4. Start the VM but will fail due to `Error: failed to generate container "xxx" spec: failed to generate spec: lstat /dev/vfio/37: no such file or directory`
1. Has workaround as @WebberHuang1118's [#6892(comment)](https://github.com/harvester/harvester/issues/6892#issuecomment-2446509516) 
   > 1. Stopping the VM
   > 1. Disabling the PCIDeviceClaim
   > 1. Waiting until the PCIDeviceClaim is removed
   > 1. Enabling the PCIDeviceClaim
   > 1. Restarting the VM

Thus decreasing the severity.

**Comment by Yu-Jack** [comment]
BTW, since the USB device path will be  also changed after rebooting. The VM won't be able to start after upgrading. 

But, it only happens that users plug in the USB device, then re-plug it in again. For example, after users plug in the USB device for the first time, the path is 002. After re-plugging it in, the path is changed to 003. Then, users decide to pass through USB devices. Next, users try to upgrade. After rebooting, the path will be changed back to 002.

However, if the users don't re-plug it in, the path is still 002 after rebooting. In this case, users can still start vm.

I had already included those limitations into [Harvester documentation](https://docs.harvesterhci.io/v1.4/advanced/addons/pcidevices#limitations). Users need to follow the description provided under the limitations to resolve this issue.

However, since the USB device path has been changed, users are unable to  remove the non-existed USB device from GUI. Instead, they can only remove it from `Edit as YAML` option. 

![image](https://github.com/user-attachments/assets/1bac5ce2-a1d6-49a4-88eb-8c347f70f60d)

![image](https://github.com/user-attachments/assets/61cef067-6fc9-46b4-93d5-dda4207246c0)

**Comment by albinsun** [comment]
Hi @Yu-Jack and @WebberHuang1118 ,
Could you help to check workaround steps? cuz I just retested it observe that PCI device disable may denined by admission webhook (Forget why I didn't hit last time :sweat_smile:).

![image](https://github.com/user-attachments/assets/c8d0276f-715f-49f4-b8fc-455f3acf0e6a)

![image](https://github.com/user-attachments/assets/ca5b4a17-3584-41c4-a0dc-c658b486187b)

**Comment by Yu-Jack** [comment]
> [!WARNING]  
> This comment workaround is for https://github.com/harvester/harvester/issues/6892#issuecomment-2458728002, **not for original issue (`[BUG] VM with assigned PCI GPU can not start after v1.3.2 to v1.4.0-rc4 upgrade due to no /dev/vfio/xx path`)**. We created a new issue from https://github.com/harvester/harvester/issues/6892#issuecomment-2458728002 in https://github.com/harvester/harvester/issues/6957. 

---

Hi @albinsun and @WebberHuang1118 I think it's a bug due to `harvesterhci.io/deviceAllocationDetails` annotation. (pcidevice wehook validates this annotation)

Although you removed the `spec.hostDevices`, the `harvesterhci.io/deviceAllocationDetails` still contained the pcidevice value. It's a new mechanism added in https://github.com/harvester/pcidevices/pull/91 after VM is stopped. 

The workaround is that:

1. Remove the annotation. Then, click "Save".
2. Detach it and disable it. Then, click "Save"

You can check my recording for more details. I think we should open a new issue to track this.

https://github.com/user-attachments/assets/db7e8788-fad8-472f-b7f9-bd0ce02b8438

**Comment by Yu-Jack** [comment]
> [!WARNING]  
> We created a new issue https://github.com/harvester/harvester/issues/6958 to track the issue from this comment. 
>
> This comment isn't related to original issue `[BUG] VM with assigned PCI GPU can not start after v1.3.2 to v1.4.0-rc4 upgrade due to no /dev/vfio/xx path`

---

BTW, I also noticed that the VM wasn't really updated yet when you detached the pcidevice from virtual machine spec. However, the disabling pcidevice request is sent. So, I'm guessing that v1.3.2 might have the same issue.

https://github.com/user-attachments/assets/0852a097-bb7c-47c6-aca9-a77392689005

**Comment by Yu-Jack** [comment]
> [!WARNING]  
> We created a new issue https://github.com/harvester/harvester/issues/6958 to track the issue from this comment. 
>
> This comment isn't related to original issue `[BUG] VM with assigned PCI GPU can not start after v1.3.2 to v1.4.0-rc4 upgrade due to no /dev/vfio/xx path`

---

Yeah, my assumption is correct. This bug has present for a while. This is a recording from v1.3.2 Harvester. I think we can enhance the process of detaching and disabling the pcidevice while editing the virtual machine.

https://github.com/user-attachments/assets/f7307de9-1f16-4b89-8314-67cd0696c3da

**Comment by WebberHuang1118** [comment]
> Hi @albinsun and @WebberHuang1118 I think it's a bug due to `harvesterhci.io/deviceAllocationDetails` annotation. (pcidevice wehook validates this annotation)
> 
> Although you removed the `spec.hostDevices`, the `harvesterhci.io/deviceAllocationDetails` still contained the pcidevice value. It's a new mechanism added in [harvester/pcidevices#91](https://github.com/harvester/pcidevices/pull/91) after VM is stopped.
> 
> The workaround is that:
> 
> 1. Remove the annotation. Then, click "Save".
> 2. Detach it and disable it. Then, click "Save"
> 
> You can check my recording for more details. I think we should open a new issue to track this.
> 
>  failed.mov

After applying the workaround mentioned by @Yu-Jack, the VM can start using the PCI device.

**Comment by Yu-Jack** [comment]
> [!WARNING]  
> We created two new issues https://github.com/harvester/harvester/issues/6957 and https://github.com/harvester/harvester/issues/6958 to track the issues from this comment. 
>
> This comment isn't related to original issue `[BUG] VM with assigned PCI GPU can not start after v1.3.2 to v1.4.0-rc4 upgrade due to no /dev/vfio/xx path`

---

TL;DR 

Just discussed with @WebberHuang1118 , there are two bugs in GUI.

1. In v1.3.2 and v1.4.0-rc4 Harvester, there is a small flaw in the GUI process of detaching and disabling pcidevice. Please check [v1.4.0 recording](https://github.com/harvester/harvester/issues/6892#issuecomment-2458803798) and [v1.3.2 recording](https://github.com/harvester/harvester/issues/6892#issuecomment-2458849222) for mode details. [Workaround for v1.4.0-rc4 Harvester is here](https://github.com/harvester/harvester/issues/6892#issuecomment-2458793241). Anyway, removing pcidevice from vm and disabling pcidevice should be split into two steps instead of one step. But, I think it's been existing from beginning .......
2. In v1.4.0-rc4 Harvester, beside removing the `spec.hostDevices`, GUI should also remove the `harvesterhci.io/deviceAllocationDetails` annotation.

**Comment by samueldewever** [comment]
We have the same issue with VMs on our 1.3.2 cluster with SR-IOV VF mappings (which use the same logic as GPU I believe). 

We have multiple VMs which can't boot if they have a SR-IOV VF mapped , same eror (/dev/vfio/xxx not found). When I disconect the PCI device, disable PCI passtrough on the VF and reenable it, I can use it again.

I am not sure if this happened after a host reload, as we experience this a lot in our labo environment and hosts get regular reboots in this environment.

The workaround is really not working when you have a lot of VMs with VF.
SUSE Support # 01559378 is opened and we have uploaded a support bundle and VM logs in the ticket. 

Might be good to confirm that its related to the same bug and see if we can have a fix in a faster release (1.4.0 or 1.4.1?)

**Comment by ibrokethecloud** [comment]
yeah this is a bug, and we will look to fix it quickly. I cannot commit an ETA right now but will be a high priority fix.

**Comment by samueldewever** [comment]
@WebberHuang1118 any update on this case? We are happy to test new version of the pcidevices once available

**Comment by WebberHuang1118** [comment]
@samueldewever 
To test the new version, you can consider updating the pcidevice daemonset as https://github.com/harvester/pcidevices/pull/97. 

However, it's not the official release, the PR will be included in the next pcidevice addon version.

**Comment by TachunLin** [comment]
We also encounter the same issue happen after upgrade from `v1.4.0` to `v1.4.1-rc1`

* The VM with GPU PCI device stuck in `Starting` state with failure
  ![Image](https://github.com/user-attachments/assets/a6ee8fab-9956-471c-8e86-4bedceffc242)
  
  ```
  	
  Pod virt-launcher-sles-gpu-r82vg
  (combined from similar events): Error: failed to generate container "9903467eeebce5040a0b3a2d88df44bc40bf409a5371b62097fce9bebcd50dee" spec: failed to generate spec: lstat /dev/vfio/34: no such file or directory
  ```
  
  https://github.com/user-attachments/assets/41430d63-a37f-407a-8fe9-b4157fbdd833


### Support bundle

[supportbundle_794c867e-54a4-48c7-9ec1-a3ae32a3e00d_2025-01-20T06-57-53Z.zip](https://github.com/user-attachments/files/18473563/supportbundle_794c867e-54a4-48c7-9ec1-a3ae32a3e00d_2025-01-20T06-57-53Z.zip)

**Comment by samueldewever** [comment]
> We also encounter the same issue happen after upgrade from `v1.4.0` to `v1.4.1-rc1`
> 
> * The VM with GPU PCI device stuck in `Starting` state with failure
>   ![Image](https://github.com/user-attachments/assets/a6ee8fab-9956-471c-8e86-4bedceffc242)
>   ```
>   	
>   Pod virt-launcher-sles-gpu-r82vg
>   (combined from similar events): Error: failed to generate container "9903467eeebce5040a0b3a2d88df44bc40bf409a5371b62097fce9bebcd50dee" spec: failed to generate spec: lstat /dev/vfio/34: no such file or directory
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   
>     
>       
>       
>   
>       vokoscreenNG-2025-01-20_14-55-10.mp4

@TachunLin this is related to the pcidevices daemonset which has a separate release cycle. (Add-on)

Could you test harvester/pcidevices#97 https://github.com/harvester/pcidevices/pull/97 ?

We still plan to test this in lab.

**Comment by TachunLin** [comment]
I think it would be similar to the https://github.com/harvester/pcidevices/pull/97 
Since the VM with GPU PCI device have been rebooted after upgrade.

And I also tried the workaround in https://github.com/harvester/harvester/issues/6892#issuecomment-2451124025

It looks like I can't disable the PCI Passthrough on the GPU device

https://github.com/user-attachments/assets/38da2ae6-6258-406c-81ae-1104e1479163

**Comment by samueldewever** [comment]
> I think it would be similar to the [harvester/pcidevices#97](https://github.com/harvester/pcidevices/pull/97) Since the VM with GPU PCI device have been rebooted after upgrade.
> 
> And I also tried the workaround in [#6892 (comment)](https://github.com/harvester/harvester/issues/6892#issuecomment-2451124025)
> 
> It looks like I can't disable the PCI Passthrough on the GPU device
> 
>  vokoscreenNG-2025-01-20_15-55-16.mp4

 @TachunLin which version of the harvester-pcidevices-controller are you running? 0.4.1?

**Comment by samueldewever** [comment]
> [@samueldewever](https://github.com/samueldewever) To test the new version, you can consider updating the pcidevice daemonset as [harvester/pcidevices#97](https://github.com/harvester/pcidevices/pull/97).
> 
> However, it's not the official release, the PR will be included in the next pcidevice addon version.

@Yu-Jack @WebberHuang1118 Any reason why this PR was not included in the latest release 0.4.1?
When is the release with this fixed planned?

**Comment by WebberHuang1118** [comment]
This PR is not contained in the pcidevices-controller v0.4.1, since it's already released.

It will be included in the next release

**Comment by TachunLin** [comment]
The `harvester-pcidevices-controller` version in the pod definition is `0.4.1`


```
# kubectl get pods harvester-pcidevices-controller-cdhjv -n harvester-system -o yaml
...
  containerStatuses:
  - containerID: containerd://b54cf584eba7acd2b0eaf4490e071f5e1c59131c274bcc1800f0ba7fc7e38e36
    image: docker.io/rancher/harvester-pcidevices:v0.4.1
    imageID: sha256:aac5e9f4806e0002b02007624334c3617126415013c4af28157a645fa2fecc70
    lastState:
      terminated:
        containerID: containerd://1c28306056f79af9390f8144516db80ff7d649632913d403185603ff31a5f345
        exitCode: 255
        finishedAt: "2025-01-20T06:24:41Z"
        reason: Unknown
        startedAt: "2025-01-18T15:16:54Z"
```

**Comment by albinsun** [comment]
Test PASS, close as fixed.


## Environment
* `harvester-v1.5.0-rc4` -> `harvester-v1.5-head`
  * Profile: 2 nodes AMD64 Bare-metal HPE ProLiant DL360 
  * ui-source: `Auto`


## Test Steps
1. <details>
   <summary> ✔ Enable <b>pcidevices-controller</b> add-on
   </summary>
   
   * Add-on
     ![Image](https://github.com/user-attachments/assets/08618da7-51da-44bf-b172-df6ac4b586f2)

   * PCI Devices
     ![Image](https://github.com/user-attachments/assets/266e1ef9-2d38-4caa-9612-bbdd88a91ed3)
   </details>

1. <details>
   <summary> ✔ Create a SLES-15.6 VM with a binding PCI passthrough GPU
   </summary>
   
   ![Image](https://github.com/user-attachments/assets/839dc9b9-0643-4fd4-957c-6bcc7e02837a)

   ![Image](https://github.com/user-attachments/assets/a63bfe54-2b11-4ddb-8a84-d94d80a17476)
   </details>
   
1. <details>
   <summary> Verify GPU works
   </summary>
   
    * Setup
      ```
      Setup Env
      # zypper install pciutils gcc git-core cmake
      # zypper addrepo https://download.opensuse.org/repositories/home:pontostroy/openSUSE_Factory/home:pontostroy.repo
      # zypper refresh
      # zypper install libOpenCL1
      # zypper addrepo https://developer.download.nvidia.com/compute/cuda/repos/sles15/x86_64/cuda-sles15.repo
      # SUSEConnect --product PackageHub/15.6/x86_64
      # zypper refresh
      # zypper install nvidia-compute-utils-G06
      # reboot

      Check setup
      # nvidia-smi
      ```
   * Check & Run CUDA Sample
      ```
      Install CUDA Toolkit
      # zypper in cuda cuda-toolkit
      
      Run CUDA Sample
      # git clone https://github.com/nvidia/cuda-samples
      # cd cuda-samples/Samples/3_CUDA_Features/cudaTensorCoreGemm
      # cmake . -DCMAKE_CUDA_COMPILER=/usr/local/cuda/bin/nvcc
      # make
      # ./cudaTensorCoreGemm
      ```
      Ref. https://github.com/NVlabs/instant-ngp/issues/1142
   * Result
     ![Image](https://github.com/user-attachments/assets/677e7fed-56a5-4d9f-a27c-809ab514fe46)
   </details>
   
1. <details>
   <summary> ✔ Upgrade Harvester to <code>v1.5-head</code>
   </summary>
   
   * Should ask to shutoff the VM since it's binding to specific node
     ![Image](https://github.com/user-attachments/assets/319c0d9b-8b61-4cb5-bf69-d95b2e6527b7)

   * Shutoff VM and start upgrade
     ![Image](https://github.com/user-attachments/assets/33e32042-8b4f-4cc3-8728-7b89b04ab054)

   * Result
     ![Image](https://github.com/user-attachments/assets/ea0d39f3-2a52-4cd1-a17d-7bf103687dcc)
   </details>


1. <details>
   <summary> ✔ Start VM and verify GPU still works
   </summary>

   * Running
     ![Image](https://github.com/user-attachments/assets/1b4b0a7a-5269-44e5-80f6-1400d344dd18)

   * GPU still works
     ![Image](https://github.com/user-attachments/assets/67cacb5a-7b43-46cc-b8a4-4750ecb81b74)
   </details>
